### ISSUE TITLE
Fix false negatives for where, is, nth-child and nth-last-child in selector-max-*

### ISSUE THREAD

**Issue by malsf21** [stylelint/stylelint#4836]
> Clearly describe the bug

As discussed in issue #4619, it seems like the `isLogicalCombination` utility doesn't cover all the available [functional pseudo-classes](https://drafts.csswg.org/selectors-4/#pseudo-classes) that provide [evaluation contexts](https://drafts.csswg.org/selectors-4/#specificity-rules); in particular, as [this demo](https://stylelint.io/demo#N4Igxg9gJgpiBcIDEAzCEAEwC+AdAdgIbz4QAuAFKugJRZ5HwAWhAzlWhHTgcQLaEyYJjHbUu9Ar3gB3EQCcYHWpMYBLdjD4AaDOO4Ni+MkwC0wtQBsoFGADcY+DBBR7OB6cbOW2Zc0ysbe0dnV31VEG0QVgBPY0IADwRwVlZI8Ah8FDUAc2TgAgwMXBB5AFdLURL4LEKi4uiYSrAyCHlTAQTTNShqjAAGOoZsEGwgA) indicates, it fails to properly peek inside of:

* `:where()`
* `:is()`
* `:nth-child()`
* `:nth-last-child()`

Which are all examples of functional pseudo-classes mentioned in the CSS W3C draft linked above. These all adversely impact the `selector-max-id` rule, among other possible rules that use the utility as well; for `selector-max-id`, it creates a false negative if an id selector is within these pseudo-classes.

In addition, peeking at the [W3Schools page for pseudo-classes](https://www.w3schools.com/css/css_pseudo_classes.asp) also lists `:lang()`, `:nth-of-type()`, and `:nth-last-of-type()` as possible options which are currently unsupported by the util, however they fall out of the direct use case with `selector-max-id`.

> Which rule, if any, is the bug related to?

Issue with `util/isLogicalCombination`; directly related to `selector-max-id`.

> What code is needed to reproduce the bug?

As indicated in the demo,

e.g.

```css
a:where(#foo) {}
a:is(em, #foo) {}
a:nth-child(even of #foo) {}
a:nth-last-child(even of #foo) {}
```

> What stylelint configuration is needed to reproduce the bug?

As indicated in the demo,

e.g.

```json
{
  "rules": {
   "selector-max-id": 0
  }
}
```

> Which version of stylelint are you using?

Latest (`13.6.1`)

> How are you running stylelint: CLI, PostCSS plugin, Node.js API?

Indifferent.

> Does the bug relate to non-standard syntax (e.g. SCSS, Less etc.)?

No.

> What did you expect to happen?

Warnings to be flagged on each use of `#foo`.

> What actually happened (e.g. what warnings or errors did you get)?

No warnings were flagged.

As @jeddy3 mentions in #4619, the solution would involve editing the following switch-case:

https://github.com/stylelint/stylelint/blob/58aab8def87fdceec4bbc4e87591892def144025/lib/utils/isLogicalCombination.js#L9-L22

If this is a good assessment of the issue, I'd be more than happy to quickly update the library and resulting tests. Let me know if I'm either missing any pseudo-classes or proposing an inefficient solution!

**Comment by jeddy3** [comment]
@malsf21 Thank you for writing this up so clearly.

> These all adversely impact the selector-max-id rule, among other possible rules that use the utility as well;

It'll impact all the `selector-max-*` rules.

> In addition, peeking at the W3Schools page for pseudo-classes also lists `:lang()`, `:nth-of-type()`, and `:nth-last-of-type()`

We can ignore these as:

- `:lang()` only expects an `<ident>` ([spec](https://drafts.csswg.org/selectors/#the-lang-pseudo))
- `:nth-of-type()` and `:nth-last-of-type()` don't expect a selector list ([spec](https://drafts.csswg.org/selectors/#the-nth-of-type-pseudo))

> Let me know if I'm either missing any pseudo-classes or proposing an inefficient solution!

We should take this opportunity to refactor the utility as it should be using [the references keywords](https://github.com/stylelint/stylelint/blob/master/lib/reference/keywordSets.js). I suggest we:

1. Pull `is`, `not`, `matches` and `has` out of [this set](https://github.com/stylelint/stylelint/blob/master/lib/reference/keywordSets.js#L223) and into a new `logicalCombinationsPseudoClasses` set.
2. Add `where` to this set.
3. Pull `nth-child` and `nth-last-child` out of [this set](https://github.com/stylelint/stylelint/blob/master/lib/reference/keywordSets.js#L210) and into a new `aNPlusBOfSNotationPseudoClasses` set.
4. Rename `isLogicalCombination` to `isContextFunctionalPseudoClass`, and replace the `switch` with requiring the two new sets and checking for a match.

At some point, we'll probably to bring all the set names up to date with the latest specifications but I think these changes will suffice for now as they're what's needed to address the false negatives.

**Comment by malsf21** [comment]
Great, thank you for outlining what needs to be changed - I'll get working on it today or tomorrow!
